### PR TITLE
JA4 for TLS and QUIC -- v9

### PR DESCRIPTION
**Pull request by satta**
Previous PR: #10341

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: https://redmine.openinfosecfoundation.org/issues/6379

Changes to previous PR:
* Rebase against current master.
* Removed raw JA4 plans from documentation.
* Use `ja4.ciphersuites.len()` instead of separate counter.
* Added comments for the reasoning behind some tests.
* Removed stowaway `#define` carried over from older code.
* Moved `JA4_HEX_LEN` into separate header to be used in detect and output code.
* Reduce number of includes in JA4 detection implementation.
* Do not allocate an empty JA3 hash when JA3 is disabled.
* Restructure some code for better readability.
* Pull out earlier feature tracking initialization into separate commit.
* Change commit message prefix.
* Moved JA3/4 enabled check into separate functions to improve readability.
* Address overflow in ALPN extension parser code.

```
SV_BRANCH=pr/1682
```

### CODE REVIEW

**Comment by satta** [comment]
Added back the ticket reference in the commit message.

File: doc/userguide/rules/ja4-keywords.rst [hunk 1]
```diff
@@ -0,0 +1,26 @@
+JA4 Keywords
+============
+
+Suricata comes with a JA4 integration (https://github.com/FoxIO-LLC/ja4). JA4,
+as part of the larger JA4+ suite of fingerprints, is used to fingerprint TLS
+clients.
+
+JA4 support must be enabled in the Suricata config file (set
+``app-layer.protocols.tls.ja4-fingerprints`` to ``yes``). If it is not
+explicitly disabled (``no``) , it will enabled if a loaded rule requires it.
```
Comment (catenacyber) [inline doc/userguide/rules/ja4-keywords.rst:10]: typo enabled/enable

File: doc/userguide/rules/ja4-keywords.rst [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+ja4.hash
+--------
+
+Match on JA4 hash (e.g. ``q13d0310h3_55b375c5d22e_cd85d2d88918``).
```
Comment (catenacyber) [inline doc/userguide/rules/ja4-keywords.rst:15]: We should mention the fact that it is a compile-time option that can be tested with `requires` keyword, cc @jasonish @jufajardini

**Review by catenacyber — CHANGES_REQUESTED** [review]
Thanks for the work Sascha :-)

- CI : needs a rebase to fix the small conflict
- Code : doing now
- Commits segmentation : nice
- Commit messages : ok for me
- Git ID set : looks fine for me
- CLA : you already contributed :-)
- Doc update : Ok
- Redmine ticket : ok
- Rustfmt : ok for ja4.rs
- Tests : Left remarks there
- Dependencies added: none

File: src/detect-tls-ja3-hash.c [hunk 3]
```diff
@@ -56,6 +56,27 @@
...
+    return -1;
+}
+
+void DetectTlsJa3HashRegister(void)
+{
+    sigmatch_table[DETECT_AL_TLS_JA3_HASH].name = "ja3.hash";
+    sigmatch_table[DETECT_AL_TLS_JA3_HASH].alias = "ja3_hash";
+    sigmatch_table[DETECT_AL_TLS_JA3_HASH].desc = "sticky buffer to match the JA3 hash buffer";
+    sigmatch_table[DETECT_AL_TLS_JA3_HASH].url = "/rules/ja3-keywords.html#ja3-hash";
+    sigmatch_table[DETECT_AL_TLS_JA3_HASH].Setup = DetectJA3SetupNoSupport;
```
Comment (catenacyber) [inline src/detect-tls-ja3-hash.c:73]: Could we minimize the changes ?
We do not need a whole new `DetectTlsJa3HashRegister` just change the `Setup` to `DetectJA3SetupNoSupport`, right ?

File: rust/src/ja4.rs [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+    }
+}
+
+// Stubs for when JA4 is disabled
+#[cfg(not(feature = "ja4"))]
+impl JA4 {
+    pub fn new() -> Self {
+        Self {
+            tls_version: None,
+            ciphersuites: Vec::with_capacity(20),
```
Comment (catenacyber) [inline rust/src/ja4.rs:53]: no need to alloc if we do not have the feature :-p

File: rust/src/ja4.rs [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+        let ja4_c_raw = format!("{}_{}", ja4_c1_raw, ja4_c2_raw);
+        sha.update(&ja4_c_raw);
+        let mut ja4_c = format!("{:x}", sha.finalize());
+        ja4_c.truncate(12);
+
+        return format!("{}_{}_{}", ja4_a, ja4_b, ja4_c);
+    }
+}
+
+pub struct SCJA4(JA4);
```
Comment (catenacyber) [inline rust/src/ja4.rs:231]: Why do we need this struct `SCJA4` ?

File: rust/src/quic/frames.rs [hunk 6]
```diff
@@ -292,11 +324,17 @@ fn parse_quic_handshake(msg: TlsMessage) -> Option<Frame> {
...
                 ja3.push(',');
                 let ciphers = ch.ciphers;
-                let extv = quic_get_tls_extensions(ch.ext, &mut ja3, true);
-                return Some(Frame::Crypto(Crypto { ciphers, extv, ja3 }));
+                let extv = quic_get_tls_extensions(ch.ext, &mut ja3, Some(&mut ja4), true);
+                return Some(Frame::Crypto(Crypto {
+                    ciphers,
+                    extv,
+                    ja3,
+                    ja4: Some(ja4),
```
Comment (catenacyber) [inline rust/src/quic/frames.rs:336]: I am afraid we will log something dummy when ja4 is not compiled in

**Comment by catenacyber** [comment]
Compiling with `--disable-ja4`, I get `"ja4": "",` in my eve.json for quic and I do not want them

File: doc/userguide/rules/ja4-keywords.rst (hunk 1 above)
Comment (satta) [reply doc/userguide/rules/ja4-keywords.rst:10]: fixed, thanks

File: doc/userguide/rules/ja4-keywords.rst (hunk 2 above)
Comment (satta) [reply doc/userguide/rules/ja4-keywords.rst:15]: Added a quick note mentioning the compile time option and the use of `requires`.

File: rust/src/ja4.rs (hunk 4 above)
Comment (satta) [reply rust/src/ja4.rs:53]: Yep, will use empty `Vec`s for these. Thanks for catching that.

File: rust/src/quic/frames.rs (hunk 6 above)
Comment (satta) [reply rust/src/quic/frames.rs:336]: Fixed.

**Comment by satta** [comment]
> Compiling with `--disable-ja4`, I get `"ja4": "",` in my eve.json for quic and I do not want them

True, I just addressed this and also double-checked that for JA3.

File: rust/src/ja4.rs (hunk 5 above)
Comment (satta) [reply rust/src/ja4.rs:231]: The idea was to have a specific type to use on the C side of things that wraps the Rust struct. But I'll check if this wrapping is really necessary or if we can just pass the Rust `JA4` struct to C directly.

File: src/detect-tls-ja3-hash.c (hunk 3 above)
Comment (satta) [reply src/detect-tls-ja3-hash.c:73]: Will do

**Comment by satta** [comment]
New PR: #10725 10725

**Pull request closed without merging**